### PR TITLE
Backup torpedo

### PR DESCRIPTION
**Pull request by raviprasadsn**
Backup Driver for Torpedo:

1. Implements backup driver
2. CRUD Interfaces for backup
3. Jenkins job run with backup driver https://jenkins.portworx.co/job/Torpedo/job/tp-dev-test/330/
4. Backup-Enabled: https://jenkins.portworx.co/job/Torpedo/job/tp-dev-test/352/
5. Backup-Disabled: https://jenkins.portworx.co/job/Torpedo/job/tp-dev-test/353/

### CODE REVIEW

File: drivers/backup/backup.go [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+	Init(schedulerDriverName string, nodeDriverName string, volumeDriverName string, token string) error
+
+	// String returns the name of this driver
+	String() string
+}
+
+// Org object interface
+type Org interface {
+	// CreateOrganization creates Organization
+	CreateOrganization(req *Request) (*api.OrganizationCreateResponse, error)
```
[the remark this answers is not in the document]
Comment (raviprasadsn) [reply drivers/backup/backup.go]: Fixed

File: drivers/backup/portworx/portworx.go [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+func (p *portworx) getSchedulePolicyManager() api.SchedulePolicyClient {
+	if p.refreshEndpoint {
+		p.setDriver(pxbServiceName, pxbNameSpace)
+	}
+	return p.schedulePolicyManager
+}
+
+func (p *portworx) CreateOrganization(req *api.OrganizationCreateRequest) (*api.OrganizationCreateResponse, error) {
+	org := p.getOrganizationManager()
+	if req != nil {
```
Comment (disrani-px) [inline drivers/backup/portworx/portworx.go]: I don't believe you need this check, the API will return an error if the request is nil.

File: tests/common.go [hunk 3]
```diff
@@ -124,6 +129,9 @@ func InitInstance() {
 
 	err = Inst().V.Init(Inst().S.String(), Inst().N.String(), token, Inst().Provisioner)
 	expect(err).NotTo(haveOccurred())
+
+	err = Inst().B.Init(Inst().S.String(), Inst().N.String(), Inst().V.String(), token)
```
Comment (disrani-px) [inline tests/common.go]: The backup server won't be running in all the tests, so this should be done only if a driver was provided.

File: drivers/backup/portworx/portworx.go [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+	if err != nil {
+		return fmt.Errorf("Error getting scheduler driver %v: %v", schedulerDriverName, err)
+	}
+
+	p.volumeDriver, err = volume.Get(volumeDriverName)
+	if err != nil {
+		return fmt.Errorf("Error getting volume driver %v: %v", volumeDriverName, err)
+	}
+
+	if err = p.setDriver(pxbServiceName, pxbNameSpace); err != nil {
```
[the remark this answers is not in the document]
Comment (raviprasadsn) [reply drivers/backup/portworx/portworx.go]: fixed

File: drivers/backup/portworx/portworx.go [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+	p.schedulePolicyManager = api.NewSchedulePolicyClient(conn)
+	p.organizationManager = api.NewOrganizationClient(conn)
+
+	logrus.Infof("Using %v as endpoint for portworx backup driver", pxEndpoint)
+
+	return err
+}
+
+func (p *portworx) GetServiceEndpoint(serviceName string, nameSpace string) (string, error) {
+	svc, err := core.Instance().GetService(serviceName, nameSpace)
```
[the remark this answers is not in the document]
Comment (raviprasadsn) [reply drivers/backup/portworx/portworx.go]: fixed

File: drivers/backup/portworx/portworx.go [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+	if err == nil && endpoint != "" {
+		if err = p.testAndSetEndpoint(endpoint); err == nil {
+			p.refreshEndpoint = false
+			return nil
+		}
+	} else if err != nil && len(node.GetWorkerNodes()) == 0 {
+		return err
+	}
+
+	p.refreshEndpoint = true
```
[the remark this answers is not in the document]
Comment (raviprasadsn) [reply drivers/backup/portworx/portworx.go]: fixed. It discovers by the service name

File: tests/common.go (hunk 3 above)
Comment (raviprasadsn) [reply tests/common.go]: Fixed.

File: drivers/backup/portworx/portworx.go (hunk 2 above)
Comment (raviprasadsn) [reply drivers/backup/portworx/portworx.go]: fixed

File: drivers/backup/portworx/portworx.go [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+	for _, n := range node.GetWorkerNodes() {
```
[the remark this answers is not in the document]
Comment (raviprasadsn) [reply drivers/backup/portworx/portworx.go]: fixed

File: drivers/backup/portworx/portworx.go [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+}
+
+func (p *portworx) String() string {
+	return driverName
+}
+
+func (p *portworx) Init(schedulerDriverName string, nodeDriverName string, volumeDriverName string, token string) error {
+	var err error
+
+	logrus.Infof("using portworx px-backup driver under scheduler: %v", schedulerDriverName)
```
[the remark this answers is not in the document]
Comment (raviprasadsn) [reply drivers/backup/portworx/portworx.go]: fixed

File: deployments/deploy-ssh.sh [hunk 9]
```diff
@@ -315,7 +325,8 @@ spec:
             "--custom-config", "$CUSTOM_APP_CONFIG_PATH",
             "--storage-upgrade-endpoint-url=$UPGRADE_ENDPOINT_URL",
             "--storage-upgrade-endpoint-version=$UPGRADE_ENDPOINT_VERSION",
-            "$APP_DESTROY_TIMEOUT_ARG"
+            "$APP_DESTROY_TIMEOUT_ARG",
+            "--backup-driver", "$BACKUP_DRIVER"
```
Comment (disrani-px) [inline deployments/deploy-ssh.sh]: Can you confirm this doesn't throw an error if the driver name is empty

File: drivers/backup/portworx/portworx.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (disrani-px) [reply drivers/backup/portworx/portworx.go]: I still see this...

File: drivers/backup/portworx/portworx.go [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+func (p *portworx) setDriver(serviceName string, nameSpace string) error {
+	var err error
+	var endpoint string
+
+	endpoint, err = p.GetServiceEndpoint(serviceName, nameSpace)
+	if err == nil && endpoint != "" {
+		if err = p.testAndSetEndpoint(endpoint); err == nil {
+			return nil
+		}
+	} else if err != nil && len(node.GetWorkerNodes()) == 0 {
```
Comment (disrani-px) [inline drivers/backup/portworx/portworx.go]: Why is this check required?

File: drivers/backup/portworx/portworx.go [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+		return err
+	}
+	return fmt.Errorf("failed to get endpoint for portworx backup driver: %v", err)
+}
+
+func (p *portworx) CreateOrganization(req *api.OrganizationCreateRequest) (*api.OrganizationCreateResponse, error) {
+	org := p.organizationManager
+	return org.Create(context.Background(), req)
```
Comment (disrani-px) [inline drivers/backup/portworx/portworx.go]:
```suggestion
	return p.organizationManager.Create(context.Background(), req)
```
Same for all the other APIs

File: deployments/deploy-ssh.sh (hunk 9 above)
Comment (raviprasadsn) [reply deployments/deploy-ssh.sh]: Run with "--backup-driver pxb"
Backup-Enabled: https://jenkins.portworx.co/job/Torpedo/job/tp-dev-test/352/. 

Run with driver name empty
Backup-Disabled: https://jenkins.portworx.co/job/Torpedo/job/tp-dev-test/353/

File: drivers/backup/portworx/portworx.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (raviprasadsn) [reply drivers/backup/portworx/portworx.go]: Fixed here https://github.com/portworx/torpedo/pull/429/commits/a8b35120d61aefff4e5ca6e662258a7230f627be

File: drivers/backup/portworx/portworx.go (hunk 10 above)
Comment (raviprasadsn) [reply drivers/backup/portworx/portworx.go]: Fixed here https://github.com/portworx/torpedo/pull/429/commits/ddfda72ea2e7507393b8960fdc10429f5fc7e5ce

File: drivers/backup/portworx/portworx.go (hunk 11 above)
Comment (raviprasadsn) [reply drivers/backup/portworx/portworx.go]: fixed here https://github.com/portworx/torpedo/pull/429/commits/a8b35120d61aefff4e5ca6e662258a7230f627be

File: drivers/backup/backup.go [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+	// Org interface
+	Org
+	// CloudCredential interface
+	CloudCredential
+	// Cluster interface
+	Cluster
+	// Backup location interface
+	BLocation
+	// Backup interface
+	Backup
```
Comment (disrani-px) [inline drivers/backup/backup.go]: Also need Restore APIs

**Pull request merged**